File: dotnet/Zirve.Sdk/Gateway/GatewayManager.cs
using System;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using Zirve.Sdk.Config;

namespace Zirve.Sdk.Gateway;

public class GatewayManager
{
    private readonly HttpClient _httpClient;
    private readonly string _url;

    public GatewayManager(HttpClient httpClient, ConfigManager configManager)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        
        var cfg = configManager.Module("gateway");
        _url = (cfg.GetValueOrDefault("url", "http://kong-kong-admin.zirve-infra.svc.cluster.local:8001") ?? "").TrimEnd('/');
    }

    private async Task<JsonElement> RequestAsync(HttpMethod method, string path, object? body = null)
    {
        var request = new HttpRequestMessage(method, $"{_url}/{path.TrimStart('/')}");
        
        if (body != null)
        {
            var json = body is string s ? s : JsonSerializer.Serialize(body);
            request.Content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");
        }

        var response = await _httpClient.SendAsync(request);
        
        if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
            return default;

        response.EnsureSuccessStatusCode();

        if (response.StatusCode == System.Net.HttpStatusCode.NoContent || response.Content.Headers.ContentLength == 0)
            return default;

        using var doc = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync());
        return doc.RootElement.Clone();
    }

    public async Task<JsonElement> AddServiceAsync(string name, string protocol, string host, int port, string path = "/")
    {
        return await RequestAsync(HttpMethod.Post, "services", new
        {
            name, protocol, host, port, path
        });
    }

    public async Task<JsonElement> AddRouteAsync(string serviceIdOrName, string name, string[] paths)
    {
        return await RequestAsync(HttpMethod.Post, $"services/{Uri.EscapeDataString(serviceIdOrName)}/routes", new
        {
            name, paths, strip_path = true
        });
    }

    public async Task<JsonElement> AddPluginAsync(string serviceIdOrName, string name, object config)
    {
        return await RequestAsync(HttpMethod.Post, $"services/{Uri.EscapeDataString(serviceIdOrName)}/plugins", new
        {
            name, config
        });
    }

    public async Task<bool> HealthAsync()
    {
        try
        {
            var response = await _httpClient.GetAsync($"{_url}/status");
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }
}

File: typescript/src/Gateway/GatewayManager.ts
/**
 * Zirve Gateway Manager — Kong
 *
 * Interacts with Kong Admin API for routing and plugins.
 */
export class GatewayManager {
  private config: Record<string, any>;
  private url: string;

  constructor(config: Record<string, any>) {
    this.config = config;
    this.url = (config.url || 'http://kong-kong-admin.zirve-infra.svc.cluster.local:8001').replace(/\/$/, '');
  }

  private async request(method: string, path: string, body?: any): Promise<any> {
    const options: RequestInit = {
      method,
      headers: { 'Content-Type': 'application/json' },
      body: body ? JSON.stringify(body) : undefined
    };

    const response = await fetch(`${this.url}/${path.replace(/^\/+/, '')}`, options);
    
    if (!response.ok) {
      if (response.status === 204 || response.status === 404) return null;
      throw new Error(`Kong API Error [${response.status}]`);
    }

    if (response.status === 204) return true;
    return await response.json();
  }

  public async addRoute(serviceIdOrName: string, name: string, paths: string[]): Promise<any> {
    return await this.request('POST', `services/${serviceIdOrName}/routes`, {
      name,
      paths,
      strip_path: true
    });
  }

  public async addService(name: string, protocol: string, host: string, port: number, path: string = '/'): Promise<any> {
    return await this.request('POST', 'services', {
      name,
      protocol,
      host,
      port,
      path
    });
  }

  public async addPlugin(serviceIdOrName: string, name: string, config: any): Promise<any> {
    return await this.request('POST', `services/${serviceIdOrName}/plugins`, {
      name,
      config
    });
  }

  public async rateLimit(serviceIdOrName: string, minute: number, hour: number): Promise<any> {
    return await this.addPlugin(serviceIdOrName, 'rate-limiting', {
      minute,
      hour,
      policy: 'local'
    });
  }

  public async health(): Promise<boolean> {
    try {
      const response = await fetch(`${this.url}/status`);
      return response.ok;
    } catch {
      return false;
    }
  }
}
